File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/provider/ListingOperation.java
package app.zxtune.fs.provider;

import android.database.Cursor;
import android.net.Uri;

import androidx.annotation.Nullable;

import java.util.Comparator;

import app.zxtune.fs.VfsDir;
import app.zxtune.fs.VfsExtensions;
import app.zxtune.fs.VfsObject;

class ListingOperation implements AsyncQueryOperation {

  private final Uri uri;
  private final Resolver resolver;
  private final ListingCursorBuilder builder;

  ListingOperation(Uri uri, Resolver resolver, SchemaSource schema) {
    this.uri = uri;
    this.resolver = resolver;
    this.builder = new ListingCursorBuilder(schema);
  }

  @SuppressWarnings("unchecked")
  @Nullable
  @Override
  public Cursor call() throws Exception {
    final VfsDir dir = maybeResolve();
    if (dir != null) {
      dir.enumerate(builder);
      return builder.getSortedResult((Comparator<VfsObject>) dir.getExtension(VfsExtensions.COMPARATOR));
    } else {
      return null;
    }
  }

  @Nullable
  private VfsDir maybeResolve() throws Exception {
    final VfsObject obj = resolver.resolve(uri);
    if (obj instanceof VfsDir) {
      return (VfsDir) obj;
    }
    return null;
  }

  @Override
  public Cursor status() {
    return builder.getStatus();
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/provider/ListingOperation.kt
package app.zxtune.fs.provider

import android.net.Uri
import app.zxtune.fs.VfsDir
import app.zxtune.fs.VfsFile
import app.zxtune.fs.comparator

internal class ListingOperation(
    private val uri: Uri,
    private val resolver: Resolver,
    private val schema: SchemaSource,
    private val callback: AsyncQueryOperation.Callback,
) : AsyncQueryOperation {

    private val builder = ListingCursorBuilder()

    override fun call() = maybeResolve()?.let { dir ->
        dir.enumerate(object : VfsDir.Visitor() {
            override fun onItemsCount(count: Int) = builder.reserve(count)
            override fun onProgressUpdate(done: Int, total: Int) {
                callback.checkForCancel()
                builder.setProgress(done, total)
                callback.onStatusChanged()
            }

            override fun onDir(dir: VfsDir) = builder.addDir(dir)
            override fun onFile(file: VfsFile) = builder.addFile(file)
        })
        callback.checkForCancel()
        builder.sort(dir.comparator).getResult(schema)
    }

    private fun maybeResolve() = resolver.resolve(uri) as? VfsDir

    override fun status() = builder.status
}
